### ISSUE TITLE
Clojurescript support

### ISSUE THREAD

**Issue by nblumoe** [acrolinx/clj-queue-by#3]
Hey! :)

I was just checking if `clj-queue-by` did support cljs, which it doesn't and that is fine. Might be worth putting this in the comparison with core.async though.

**Comment by ska2342** [comment]
Do you have an idea what's blocking it? Is it the important `clojure.lang.PersistentQueue` or is it just the `java.io.Writer` which is used only to be able to print the thing nicely?

**Comment by nblumoe** [comment]
The two usages you mentioned and then interfaces in https://github.com/acrolinx/clj-queue-by/blob/cabeb01c0f06d06b3f118b79b9205355e85e0ae4/src/com/acrolinx/clj_queue_by.clj#L157

What are your thoughts about this? Seems like it could be made a cljc rather easily.

**Comment by ska2342** [comment]
PR welcome! You'd need to sign a contributor's agreement, though. I can arrange that to be sent to you. Doesn't require snail mail.

**Comment by nblumoe** [comment]
I started sorting out the namespace and interface differences in #4. Shouldn't be a big deal finishing and cleaning that up.
While working on it I had to realise though, that the lib uses Refs, which aren't available in CLJS. So we would need to address that. Do you have any preference for how to do that? Should we pull the code apart and use atoms on CLJS and keep Refs in CLJ?

**Comment by ska2342** [comment]
The Refs are actually important in the current impl. We're updating the queue and the internal index used for ordering items both atomically. If we can't use Refs, we need to re-do the internal datastructures.

**Comment by ska2342** [comment]
@nblumoe, the index is only used to sort items internally with respect to the order they came in. It may be possible to remove it completely and work with the current count of the queue. Need to think about this.

**Comment by nblumoe** [comment]
My thoughts about `the-index`:

1. It seems to me that using the count of the queue wouldn't be enough to keep the current behaviour. Let's say we push a new item which gets the id 4 because the current queue count is 3. Now we pop two items and then push a new item. This would now get id 3 because the queue count is 2. Thus we broke the order for the last two items pushed. In a similar way, we could easily get conflicting ids.
2. Would it be an option to retrieve the current max. id when pushing an item? That would make pushing more expensive though due to the required ad-hoc lookup. As an optimization, it should be enough to look at the last item in each queue (in `queued`, not `selected`) if I correctly understand the guarantees of the current implementation.
3. Would it be an option to include `the-index` in `the-q`? This way we would only need an atomic update on `the-q` and don't bother coordinating two mutable storage locations.
4. If we can get rid of having to coordinate changes to `the-index` and `the-q`, would there still be a need for using Refs on the JVM or could we move everything to atoms for both runtime environments (CLJ & CLJS)?

**Comment by ska2342** [comment]
ad 1: You're right. Just the count doesn't work.

ad 2: Would need some trying out how bad the impact is. I don't think that it's too serious.

ad 3: Yes, it would be an option to store the index somehow in the queue. Maybe with a dedicated, namespaced key which get's ignored on reading anything from the queue?

ad 4: That's the whole idea of getting rid of the index. No coordination :arrow_forward: no need for Refs, no Refs :arrow_forward: CLJS support is possible.

**Comment by nblumoe** [comment]
Ok great, will continue by putting the index into the queue and using Atoms instead of Refs for both runtimes.

**Comment by ska2342** [comment]
We'll have to re-think the built-in stress test, too. It wasn't the brightest idea to begin with, but I wanted to see that the queue didn't break in a multi-threaded scenario.

**Comment by nblumoe** [comment]
I updated the PR #4  with what we discussed above. I appreciate your feedback and thoughts on the code!

**Comment by ska2342** [comment]
You work on this is highly appreciated. I am skeptical about some reads outside of the safety of transactions (Refs) or compare-and-set (Atom), though. The way I understand it, there can be race-conditions.

**Comment by nblumoe** [comment]
Thanks, you were right about those issues, thanks for the review!

Seems like you reviewed an already outdated state at that point though. Two of the issues had already been solved, as they also surfaced from the stress tests. I fixed the third issue you raised (checking queue size), thanks for catching that!

I also elaborated a bit more and suggestions (not quite satisfying) solutions on another, related issue: https://github.com/acrolinx/clj-queue-by/pull/4#discussion_r280305660
